File: stacks and queues/907. Sum of Subarray Minimums.ts
`Given an array of integers arr, find the sum of min(b), where b ranges over every (contiguous) subarray of arr. Since the answer may be large, return the answer modulo 109 + 7.

 

Example 1:

Input: arr = [3,1,2,4]
Output: 17
Explanation: 
Subarrays are [3], [1], [2], [4], [3,1], [1,2], [2,4], [3,1,2], [1,2,4], [3,1,2,4]. 
Minimums are 3, 1, 2, 4, 1, 1, 2, 1, 1, 1.
Sum is 17.
Example 2:

Input: arr = [11,81,94,43,3]
Output: 444
`;

const sumSubarrayMins = (arr: number[]) => {
  let left_to_right: number[] = new Array(arr.length).fill(0);

  let stack: number[] = [];

  for (let i = 0; i < arr.length; i++) {
    while (stack.length && arr[stack[stack.length - 1]] > arr[i]) {
      // if the current element is smaller than the top of the stack, pop the stack
      left_to_right[stack.pop()!] = i;
    }
    stack.push(arr[i]);
  }

  // now right ot left
  let right_to_left: number[] = new Array(arr.length).fill(-1);
  stack = []; // reset the stack
  // going from right to left to find the left boundry
  for (let i = arr.length - 1; i >= 0; i--) {
    while (stack.length && arr[stack[stack.length - 1]] >= arr[i]) {
      // if the current element is smaller than the top of the stack, pop the stack
      right_to_left[stack.pop()!] = i;
    }
    stack.push(arr[i]);
  }

  let res = 0;
  for (let i = 0; i < arr.length; i++) {
    let [left, right] = [right_to_left[i], left_to_right[i]];
    res += (right - i) * (i - left) * arr[i];
    res %= 1000000007;
  }
  return res;
};

console.log(sumSubarrayMins([3, 1, 2, 4]));

File: stacks and queues/907. Sum of Subarray Minimums.py

from typing import List
class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        # Time: O(n) | Space: O(n)
        MOD=10**9+7

       # right boundary of i, where arr[i] is the minima.
        leftToRight = [len(arr)] * len(arr)
        
        # going from left to right to find the right boundary.
        stack = []
        for i in range(len(arr)):
            while stack and arr[i] <= arr[stack[-1]]:
                leftToRight[stack.pop()] = i
            stack.append(i)
        
        # Left boundary of i, where arr[i] is the minima.
        rightToLeft = [-1] * len(arr)
        # going from right to left to find the left boundary.
        stack = [] # resetting the stack 
        for i in reversed(range(len(arr))):
            while stack and arr[i] < arr[stack[-1]]:
                # if the current element is smaller than the top of the stack, pop the stack
                rightToLeft[stack.pop()] = i 
            stack.append(i)

        res = 0
        
        for idx in range(len(arr)):
            left, right = rightToLeft[idx], leftToRight[idx]
            res += (right-idx) * (idx-left) * arr[idx]
            res %= MOD
            
        return res

 

    
print(Solution().sumSubarrayMins([3,1,2,4]))